agent/backend/otel/otel.go

```diff
@@ -0,0 +1,237 @@
+package otel
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/go-cmd/cmd"
+	"github.com/orb-community/orb/agent/backend"
+	"github.com/orb-community/orb/agent/config"
+	"github.com/orb-community/orb/agent/otel"
+	"github.com/orb-community/orb/agent/otel/otlpmqttexporter"
+	"github.com/orb-community/orb/agent/policies"
+	"go.opentelemetry.io/collector/exporter"
+	"go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
+	"os"
+	"time"
+)
+
+var _ backend.Backend = (*openTelemetryBackend)(nil)
+
+const DefaultPath = "/usr/local/bin/otelcol-contrib"
+
+type openTelemetryBackend struct {
+	logger    *zap.Logger
+	startTime time.Time
+
+	//policies
+	policyRepo            policies.PolicyRepo
+	policyConfigDirectory string
+	agentTags             map[string]string
+
+	// Context for controlling the context cancellation
+	mainContext        context.Context
+	runningCollectors  map[string]runningPolicy
+	mainCancelFunction context.CancelFunc
+
+	// MQTT Config for OTEL MQTT Exporter
+	mqttConfig config.MQTTConfig
+	mqttClient *mqtt.Client
+
+	otlpMetricsTopic string
+	otlpTracesTopic  string
+	otlpLogsTopic    string
+	otelReceiverTaps []string
+	otelCurrVersion  string
+
+	otelReceiverHost string
+	otelReceiverPort int
+
+	metricsReceiver    receiver.Metrics
+	metricsExporter    exporter.Metrics
+	otelExecutablePath string
+
+	//tracesReceiver  receiver.Traces
+	//tracesExporter  exporter.Traces
+	//logsReceiver    receiver.Logs
+	//logsExporter    exporter.Logs
+}
+
+// Configure initializes the backend with the given configuration
+func (o *openTelemetryBackend) Configure(logger *zap.Logger, repo policies.PolicyRepo,
+	config map[string]string, otelConfig map[string]interface{}) error {
+	o.logger = logger
+	o.logger.Info("configuring OpenTelemetry backend")
+	o.policyRepo = repo
+	var err error
+	o.otelReceiverTaps = []string{"otelcol-contrib", "receivers", "processors", "extensions"}
+	o.policyConfigDirectory, err = os.MkdirTemp("", "otel-policies")
+	if path, ok := config["binary"]; ok {
+		o.otelExecutablePath = path
+	} else {
+		o.otelExecutablePath = DefaultPath
+	}
+	if err != nil {
+		o.logger.Error("failed to create temporary directory for policy configs", zap.Error(err))
+		return err
+	}
+	if agentTags, ok := otelConfig["agent_tags"]; ok {
+		o.agentTags = agentTags.(map[string]string)
+	}
+	for k, v := range otelConfig {
+		switch k {
+		case "Port":
+			o.otelReceiverPort = v.(int)
+		}
+	}
+	if o.otelReceiverPort == 0 {
+		// TODO use default values if not configured in YAML
+		o.otelReceiverHost = "localhost"
+		o.otelReceiverPort = 4317
+	}
+
+	return nil
+}
+
+func (o *openTelemetryBackend) GetInitialState() backend.RunningStatus {
+	return backend.Waiting
```

Comment:
because otel will wait until a policy comes to start actually scraping data, not like diode or pktvisor which they have a default policy of their own.